### ISSUE TITLE
TaxBrain suggestions

### ISSUE THREAD

**Issue by martinholmer** [OpenSourcePolicyCenter/webapp-public#151]
I have recently started using TaxBrain and am very impressed.  The user interface is clean and handsome, and the speed at which the tax calculations are done is truly impressive.  And it is cute that TaxBrain winks at me while it is thinking.  So, congratulations to T.J. and all the others who have contributed to the development of TaxBrain.

But, as impressive as it is now, TaxBrain can be made even better.

I would like to raise two related issues that would allow TaxBrain to provide policy analysis capabilities that are already built into the underlying Tax-Calculator, but that TaxBrain currently does not offer.  The first has to do with the Start Year and the second has to do with multiple-year-reform specification.  These two issues are interrelated, so please bear with me as I explain them one at a time.

**Start Year**

It seems to me that the Start Year combines two distinct concepts: the first year policy parameters can be changed and the first year of the ten-year budget window for which output is displayed.  I'm not sure what terms best describe these two concepts, so I'm hoping for some discussion about what terms should be used in TaxBrain.  But, in order to explain my suggestions, I need to use precise language.  In this discussion I use the following terms for these two concepts:

**First Reform Year** is the year for which parameter values are shown in the boxes.  I suggest, for reasons that will become apparent below, that we limit the range of first-reform-year values to be from 2013 (the first year for which Tax-Calculator has tax policy parameters) to 2016 (the last year for which Tax-Calculator has known, rather than projected, tax policy parameters).

**First Budget Year** is the first of ten years for which we show the basic tax output.  Obviously, the chosen first-budget-year value must be no less than the chosen first-reform-year value and no greater than 2017 (because, at the moment, Tax-Calculator cannot project beyond 2026).

**Multiple-Year-Reform Specification**

Tax-Calculator has the ability to handle complex multiple-year reforms that cannot easily be specified in TaxBrain.  Let me give an example.  Suppose I want to specify a reform in which the "Maximum Taxable Earnings for Social Security" (MTE) parameter is raised to $300,000 in 2020, indexed thereafter, then raised to $400,000 in 2024, and indexed thereafter.  (This example could use any indexed tax policy parameter; there is nothing special about the use of MTE parameter.)  You can see right away that the problem is a TaxBrain user has no idea about what to specify for the MTE value in the years between the first reform year and 2020 or for the years 2021-2023.  Now, it is possible for the person using TaxBrain to download the Tax-Calculator source code from Git Hub, find the default price-inflation and wage-growth rates in the Policy class, get out a calculator, do the indexing calculations by hand, and then enter in the TaxBrain box the calculated MTE values for all the intermediate years (assuming there is enough space in the box to do that).  But, I think, we can all agree that this is far from an ideal TaxBrain user experience.

There is an easy way to avoid this problem.  Here is how I suggest specifying this example reform in the TaxBrain MTE box, under the assumption that the first-reform-year value is 2016 (and therefore, that the first value shown in the box is 118500):

```
118500,*,*,*,300000,*,*,*,400000
```

The ```*``` character indicates that the indexing rule for that policy parameter should be used to compute that year's parameter value.

It would be straight forward for TaxBrain to be able to read that string in the MTE box and translate it into a reform that can be passed to the Policy class implement_reform() method in the Tax-Calculator.

**Conclusion**

If you have read this far, I hope you will comment on these TaxBrain suggestions and offer ideas that would be better than what I've suggested.

And I also hope that some of you are on the edge of your seat saying "But wait, Martin, you've overlooked another problem!"  Indeed there is another issue, but it is less important to resolve than the two discussed above.  It has to do with how to tell TaxBrain in what year to change a parameter's indexing rule.  Right now, there seems to be no way to say, in the context of the example above, leave the 2024 MTE value of $400,000 unchanged in subsequent years (that is, change the MTE parameter from being indexed before 2024 to being not indexed in years after 2024).  There is a way to specify such a reform in Tax-Calculator, but the current version of TaxBrain does not seem to have the same capability.  If you have ideas about how to handle this third issue, by all means, share those ideas in the discussion of this issue.  If not, I'm happy to leave this third issue for later.

cc @talumbau @zrisher @mistakevin @johneo

**Comment by MattHJensen** [comment]
@martinholmer, the asterisk approach is very elegant and I like it a lot. Splitting the first reform year and first budget year also make sense, although I don't like having to give the user another option to understand and fiddle with. I suppose it won't be too bad though because most users will want to stick with the defaults. 

What do others think?

@martinholmer, as to the smaller issue you raised at the end of your comment:

> "Right now, there seems to be no way to say, in the context of the example above, leave the 2024 MTE value of $400,000 unchanged in subsequent years (that is, change the MTE parameter from being indexed before 2024 to being not indexed in years after 2024)."

I'll second that it is not as urgent as the other issues. In fact, there is currently a way to do this, although it is somewhat inelegant:

```118500,*,*,*,300000,*,*,*,400000, 400000, 400000```

cc @feenberg, @talumbau, @zrisher, @mistakevin, @johneo

**Comment by martinholmer** [comment]
Matt said:
> @martinholmer, as to the smaller issue you raised at the end of your comment:
>
>> Right now, there seems to be no way to say, in the context of the example above, leave the 2024 MTE value of $400,000 unchanged in subsequent years (that is, change the MTE parameter from being indexed before 2024 to being not indexed in years after 2024).
>
> I'll second that it is not as urgent as the other issues. In fact, there is currently a way to do this, although it is somewhat inelegant:
```
118500,*,*,*,300000,*,*,*,400000, 400000, 400000
```

Matt,  Good point about how to turn off the indexing of a parameter in some future year.
But is there a way to go the other way?  Can a not-indexed parameter be turned into an indexed parameter in some future year?  But again, these are minor issues relative to the main points being discussed in this issue.

cc @MattHJensen @feenberg, @talumbau, @zrisher, @mistakevin, @johneo

**Comment by martinholmer** [comment]
As I continue to use TaxBrain, I find it annoying that I have to type in the default value of a policy parameter for the start year.  When I start to type the default value in the text box disappears (as it should), but the default-value message just below the text box does not immediately appear until I tab out of the text box.  This leaves me, and other TaxBrain users with less than perfect short-memory recall, wondering what they should type.  If it is a reform value, I have it in mind.  But if I am specifying a reform in which the parameter does not change until after the start year, then the current version of TaxBrain requires me to retype the start year default (i.e., current-law policy) value.  This is inconvenient and error-prone and a completely unnecessary burden on the TaxBrain user.

Consider a situation where the start year is 2016 (because some policy parameters are changing in that year) but the social security maximum taxable earnings (MTE) parameter is not scheduled to change until 2018 when it will rise to $300,000. Why not allow this way of specifying this change in MTE?
```
*,*,300000
```

In other words, I'm requesting the same TaxBrain enhancement as discussed above in this webapp-public issue #151, except that the ability of TaxBrain users to use the ```*``` character is extended to include the start year.

cc @MattHJensen @talumbau @mistakevin @feenberg @zrisher,

**Comment by MattHJensen** [comment]
> In other words, I'm requesting the same TaxBrain enhancement as discussed above in this webapp-public issue #151, except that the ability of TaxBrain users to use the * character is extended to include the start year.

@martinholmer, that makes a great deal of sense. 

This request would also assist @martinholmer's TaxBrain / Tax-Calculator Selenium comparisons in https://github.com/open-source-economics/Tax-Calculator/tree/master/taxcalc/taxbrain. 

@talumbau, do you, @mistakevin, or someone else from your team have time to tackle this in the next week or two? I think this ought to be a higher priority than displaying the 'params-used' on input and output pages given the interaction with testing.

**Comment by talumbau** [comment]
OK, I will take a crack at this. I think @mistakevin will be back online soon.There is some default "format checking" for the kinds of input fields we use, so if our inputs fields reject the asterisk symbol, we can just make our own custom input fields. assigning to myself.

**Comment by martinholmer** [comment]
T.J. said:
> OK, I will take a crack at this. I think @mistakevin will be back online soon.There is some default "format checking" for the kinds of input fields we use, so if our inputs fields reject the asterisk symbol, we can just make our own custom input fields. assigning to myself.

My experience is that the current version of TaxBrain does **not** reject something like this:
```
*,*,400000
```
TaxBrain accepts this input string and proceeds to produce nonsense results.

@talumbau @MattHJensen

**Comment by talumbau** [comment]
oh this is good to know. thanks @martinholmer

**Comment by martinholmer** [comment]
Recent revisions to TaxBrain logic have implemented the `*,*,*,500000` style of reform, so this issue is being closed.

**Comment by martinholmer** [comment]
The delayed-reform TaxBrain enhancement (using the `*` notation) that was suggested in issue #151 (which was opened on February 12, 2016) is still not working correctly.  So, this issue has been reopened and will be closed when TaxBrain can handle delayed reforms for all policy parameters.  This is not a minor enhancement, but rather brings to TaxBrain essential capabilities that have been in Tax-Calculator from the beginning.  So this enhancement should be a high priority.

@MattHJensen @talumbau

**Comment by martinholmer** [comment]
As far as I can tell from my selenium testing of TaxBrain, the delayed-reform TaxBrain enhancement (using the * notation) that was suggested in issue #151 is now working correctly.  At least my tests are not generating any "500 Internal Server Error!" messages.

@MattHJensen @talumbau
